File: widgetssdk/src/main/java/com/glia/widgets/core/callvisualizer/domain/VisitorCodeRepository.java
package com.glia.widgets.core.callvisualizer.domain;

import com.glia.androidsdk.GliaException;
import com.glia.androidsdk.RequestCallback;
import com.glia.androidsdk.omnibrowse.VisitorCode;
import com.glia.widgets.di.GliaCore;

public class VisitorCodeRepository {
    private final GliaCore gliaCore;

    public VisitorCodeRepository(GliaCore gliaCore) {
        this.gliaCore = gliaCore;
    }

    public void getVisitorCode(RequestCallback<VisitorCode> callback) {
        if (!gliaCore.isInitialized()) {
            throw new GliaException("Widgets SDK is not initialized", GliaException.Cause.INVALID_INPUT);
        }
        gliaCore.getCallVisualizer().getVisitorCode(callback);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/internal/callvisualizer/domain/VisitorCodeRepository.kt
package com.glia.widgets.internal.callvisualizer.domain

import com.glia.androidsdk.GliaException
import com.glia.androidsdk.omnibrowse.VisitorCode
import com.glia.widgets.di.GliaCore
import io.reactivex.rxjava3.core.Observable

internal class VisitorCodeRepository(private val gliaCore: GliaCore) {

    fun getVisitorCode(): Observable<VisitorCode> {
        if (!gliaCore.isInitialized) {
            return Observable.error(
                GliaException(
                    "Widgets SDK is not initialized",
                    GliaException.Cause.INVALID_INPUT
                )
            )
        }
        return Observable.create { emitter ->
            gliaCore.callVisualizer.getVisitorCode { response, exception ->
                if (response != null && response.code != null) {
                    emitter.onNext(response)
                } else if (exception != null) {
                    emitter.onError(exception)
                } else {
                    emitter.onError(IllegalArgumentException("No value presented"))
                }
                emitter.onComplete()
            }
        }
    }
}
